ouroboros-network/src/Ouroboros/Network/PeerSelection/Churn.hs

```diff
@@ -61,180 +77,369 @@ peerChurnGovernor tracer deadlineChurnInterval bulkChurnInterval psOverallTimeou
   (mode, ubp) <- atomically ((,) <$> updateChurnMode
                                  <*> getUseBootstrapPeers)
   atomically $ do
-    increaseActivePeers mode
-    increaseEstablishedPeers mode ubp
+    modifyTVar peerSelectionVar ( increaseActivePeers mode
+                                . increaseEstablishedPeers mode ubp )
   endTs0 <- getMonotonicTime
-  fuzzyDelay inRng (endTs0 `diffTime` startTs0) >>= go
+  fuzzyDelay inRng (endTs0 `diffTime` startTs0) >>= churnLoop
 
   where
 
     updateChurnMode :: STM m ChurnMode
     updateChurnMode = do
         fm <- getFetchMode
         let mode = case fm of
-                        FetchModeDeadline -> ChurnModeNormal
-                        FetchModeBulkSync -> ChurnModeBulkSync
+                     FetchModeDeadline -> ChurnModeNormal
+                     FetchModeBulkSync -> ChurnModeBulkSync
         writeTVar churnModeVar mode
         return mode
 
+    -- | Update the targets to a given value, and block until they are reached.
+    -- The time we are blocked is limited by a timeout.
+    --
+    updateTargets
+      :: [ChurnAction]
+      -- ^ churn actions for tracing
+      -> DiffTime
+      -- ^ timeout
+      -> ModifyPeerSelectionTargets
+      -- ^ update counters function
+      -> CheckPeerSelectionCounters
+      -- ^ check counters
+      -> m ()
+    updateTargets churnActions timeoutDelay modifyTargets checkCounters = do
+      -- update targets, and return the new targets
+      targets <- atomically $ stateTVar peerSelectionVar ((\a -> (a, a)) . modifyTargets)
+
+      -- create timeout and block on counters
+      bracketOnError (registerDelayCancellable timeoutDelay)
```

Comment:
I was wondering if bracket can be used here instead to slightly simply things. We could then avoid calling cancelTimeout in the true branch below, but then the question is what would happen in the finalizer if timeout expired, ie. we called cancelTimeout on it. It would be nice if it was silently ignored ;)

Reply:
Our `cancelTimeout` is implemented in terms of [`unregisterTimeout`](https://hackage.haskell.org/package/base-4.19.1.0/docs/GHC-Event.html#v:unregisterTimeout) which does a lookup in a shared map (which should be cheap).